Improve sampledata downloadfromsource api

jcfr opened the pull request:
This changes are first step toward improving testing infrastructure discussed in https://discourse.slicer.org/t/improving-testing-data-management-for-self-test/5014/6

pieper commented:
LGTM 👍

lassoan commented on Applications/SlicerApp/Testing/Python/JRC2013Vis.py:
```diff
@@ -319,19 +304,15 @@ def test_Part2Head(self):
-    downloads = (
-        ('http://slicer.kitware.com/midas3/download?items=8609', '3DHeadData.mrb', slicer.util.loadScene),
+    from SampleData import SampleDataSource, SampleDataLogic
+    dataSources = [
```
Most of the time we download a single data source, so creating a list of sources is unnecessary complication. Could we change the code to this simpler pattern?

```
from SampleData import SampleDataSource, SampleDataLogic
SampleDataLogic().downloadFromSource(SampleDataSource(
  fileNames='3DHeadData.mrb', 
  uris='http://slicer.kitware.com/midas3/download?items=8609', 
  loadFileType='SceneFile'))
```

jcfr replied:
Agreed. Just wanted to get some feedback ... before simplifying.

What about that then:

```
SampleDataLogic().download(
  fileNames='3DHeadData.mrb', 
  uris='http://slicer.kitware.com/midas3/download?items=8609', 
  loadFileType='SceneFile')
```

jcfr replied:
Also we call the function `downloadFromSource` (or `download`) but most of the time it is loaded in Slicer.

Should instead of the method be named
* `SampleData.loadFromSource` or `SampleData.downloadAndLoadFromSource`
* `SampleData.load` or `SampleData.downloadAndLoad`

lassoan replied:
If the input of the method is not a `DataSource` object then we don't need to include that in the method name. So, `SampleData.load` would work well.

Probably `SampleData.loadData` or ``SampleData.loadFromURL` would be even better because it would not suggest that the input is some sample data.

`SampleData.load` could be used for loading sample data, such as `SampleData.load('MRHead')`.

lassoan replied:
What do you think about specifying a helper function at the module level to further simplify syntax?

```
import SampleData
SampleData.loadFromURL(…)
```

jcfr replied:
Or we could have methods like this directly in SampleData:

```
def loadFromURL(...):
  return SampleDataLogic().loadFromURL()

def downloadMRHead():
  return SampleDataLogic().downloadMRHead()

[...]
```

jcfr replied:
That said, I just realized that `downloadMRHead` and a like are deprecated. And that the new modern approach is to use `downloadSample("MRHead")`, it means your suggestion would work very well

lassoan commented on Applications/SlicerApp/Testing/Python/JRC2013Vis.py:
```diff
@@ -476,19 +453,15 @@ def test_Part4Lung(self):
+      SampleDataSource(
+        fileNames='LungData.mrb',
+        uris='http://slicer.kitware.com/midas3/download?items=8612',
+        loadFileType='SceneFile'
```
I think we already have a mechanism to guess based on file extension. Could we use that by default and require specifying loadFileType only if the default is not good?

jcfr commented:
@lassoan I reviewed the approach. Note that I will squash the last commits before integrating ... 

In a nutshell:

* add support for guessing loadNodeTypes ... it is still suboptimal because I am not using any Slicer logic. 

* renamed `downloadFromDataSource` to `downloadFromURL`

* added convenience method directly into SampleData

* updated code to make use of the convenience methods. It now looks like this:
```
import SampleData
SampleData.downloadFromURL(...)
mrhead = SampleData.downloadSample('MRHead')[0]
```

jcfr commented on Applications/SlicerApp/Testing/Python/SlicerBoundsTest.py:
```diff
@@ -71,8 +71,8 @@ def test_Volume(self):
-    from SampleData import SampleDataLogic
-    volumeNode = SampleDataLogic().downloadAbdominalCTVolume()
+    import SampleData
+    volumeNode = SampleData.downloadSample('CTA abdomen\n(Panoramix)')[0]
```
@lassoan  I don't link this name including a newline ... I think we should have a SampleName and SampleDescription.

jcfr commented on Modules/Loadable/Markups/Testing/Python/NeurosurgicalPlanningTutorialMarkupsSelfTest.py:
```diff
@@ -143,8 +143,8 @@ def run(self):
-    from SampleData import SampleDataLogic
-    baselineVolume = SampleDataLogic().downloadWhiteMatterExplorationBaselineVolume()
+    import SampleData
+    baselineVolume = SampleData.downloadSample('BaselineVolume')[0]
```
This SampleName is not very explicit .. should we rename it ?
